### PR TITLE
Add parameters to FineTune

### PR DESCRIPTION
**Pull request by lucas-aixplain**
Add prompt, hyperparameters and parameter-efficient fine-tuning (peft) parameters to FineTune method.
Also, it validates the prompt format to correctly point to given datasets.

Tasks: M-5366099453, M-5470304031 and M-5470353508.

### CODE REVIEW

File: aixplain/factories/finetune_factory/__init__.py [hunk 1]
```diff
@@ -78,24 +127,40 @@ def create(
...
+                    for dataset in dataset_list
+                ],
+                "sourceModelId": model.id,
+            }
+            parameters = {}
+            if prompt is not None:
+                parameters["prompt"] = prompt
+            if hyperparameters is not None:
+                parameters["hyperparameters"] = hyperparameters.to_dict()
+            if hyperparameters is not None:
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:147]: if peft is not None?

File: aixplain/modules/finetune.py [hunk 2]
```diff
@@ -87,6 +88,7 @@ def __init__(
         self.cost = cost
         self.train_percentage = train_percentage
         self.dev_percentage = dev_percentage
+        self.parameters = parameters
```
Comment (thiago-aixplain) [inline aixplain/modules/finetune.py:91]: Finetune should have a Optional[Hyperparameter] and Optional[Peft] attributes, not?

This "parameters" seems wrong.

File: aixplain/factories/finetune_factory/__init__.py [hunk 3]
```diff
@@ -78,24 +127,40 @@ def create(
...
             cost = cls._create_cost_from_response(resp)
-            return Finetune(name, dataset_list, model, cost, train_percentage=train_percentage, dev_percentage=dev_percentage)
+            return Finetune(
+                name,
+                dataset_list,
+                model,
+                cost,
+                train_percentage=train_percentage,
+                dev_percentage=dev_percentage,
+                parameters=parameters,
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:162]: should be replaced by a Optional[Hyperparameter] and Optional[Peft]

File: aixplain/factories/finetune_factory/__init__.py [hunk 4]
```diff
@@ -57,9 +60,55 @@ def _create_cost_from_response(cls, response: Dict) -> FinetuneCost:
         """
         return FinetuneCost(response["trainingCost"], response["inferenceCost"], response["hostingCost"])
 
+    @classmethod
+    def _get_data_list(cls, dataset: Dataset):
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:64]: Do not make this a classmethod. Create a separate script for this method.

File: aixplain/factories/finetune_factory/__init__.py [hunk 5]
```diff
@@ -57,9 +60,55 @@ def _create_cost_from_response(cls, response: Dict) -> FinetuneCost:
...
+        flatten_target_values = [item for sublist in list(dataset.target_data.values()) for item in sublist]
+        data_list = list(dataset.source_data.values()) + flatten_target_values
+        return data_list
+
+    @classmethod
+    def _validate_prompt(cls, prompt: Text, dataset_list: List[Dataset]) -> Text:
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:70]: Do not make this a classmethod. Create a separate script for this method together with the previous `_get_data_list`.

**Review by thiago-aixplain — CHANGES_REQUESTED** [review]
Besides the local comment: Functional tests are missing.

File: aixplain/modules/finetune.py (hunk 2 above)
Comment (lucas-aixplain) [reply aixplain/modules/finetune.py:91]: The Optional[Hyperparameter] and Optional[Peft] parameters are in the [FinetuneFactory,create method](https://github.com/aixplain/aiXplain/blob/M-5470304031-addFinetunePromptParameters/aixplain/factories/finetune_factory/__init__.py#L102). The above class is only created by FinetuneFactory and it is used to call the [sdk/finetune endpoint](https://github.com/aixplain/aiXplain/blob/M-5470304031-addFinetunePromptParameters/aixplain/modules/finetune.py#L123) from backend, so we can pass directly the parameters dictionary.

File: aixplain/factories/finetune_factory/__init__.py [hunk 6]
```diff
@@ -59,17 +62,27 @@ def _create_cost_from_response(cls, response: Dict) -> FinetuneCost:
 
     @classmethod
     def create(
-        cls, name: Text, dataset_list: List[Dataset], model: Model, train_percentage: float = 100, dev_percentage: float = 0
+        cls,
+        name: Text,
+        dataset_list: List[Dataset],
+        model: Model,
+        prompt: Optional[Text] = None,
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:69]: Should the name be `prompt_template`?

File: aixplain/factories/finetune_factory/__init__.py [hunk 7]
```diff
@@ -59,17 +62,27 @@ def _create_cost_from_response(cls, response: Dict) -> FinetuneCost:
...
+        train_percentage: Optional[float] = 100,
+        dev_percentage: Optional[float] = 0,
     ) -> Finetune:
         """Create a Finetune object with the provided information.
 
         Args:
             name (Text): Name of the Finetune.
             dataset_list (List[Dataset]): List of Datasets to be used for fine-tuning.
             model (Model): Model to be fine-tuned.
+            prompt (Text, optional): Fine-tuning prompt. Defaults to None.
```
Comment (thiago-aixplain) [inline aixplain/factories/finetune_factory/__init__.py:81]: Explain the template format here in the documentation

File: aixplain/modules/finetune/__init__.py [hunk 8]
```diff
@@ -47,6 +50,9 @@ class Finetune(Asset):
         version (Text): Version of the FineTune.
         train_percentage (float): Percentage of training samples.
         dev_percentage (float): Percentage of development samples.
+        prompt (Text): Fine-tuning prompt.
```
Comment (thiago-aixplain) [inline aixplain/modules/finetune/__init__.py:53]: maybe `prompt_template`

**Review by thiago-aixplain — APPROVED** [review]
Just add the functional tests and run them before merging.

**Pull request merged**